adapters/zio-http/src/main/scala/caliban/ZHTTPAdapter.scala

```diff
@@ -0,0 +1,240 @@
+package caliban
+
+import zio._
+import zio.clock.Clock
+import zio.duration._
+import zio.stream._
+
+import caliban.{ GraphQLInterpreter, GraphQLRequest, GraphQLResponse }
+import caliban.ResponseValue.{ ObjectValue, StreamValue }
+import caliban.ResponseValue
+import caliban.Value.NullValue
+import caliban.execution.QueryExecution
+import io.circe._
+import io.circe.parser._
+import io.circe.syntax._
+import zhttp.http._
+import zhttp.socket.SocketApp
+import zhttp.socket.WebSocketFrame.Text
+import zhttp.socket._
+
+object ZHttpAdapter {
+  case class GraphQLWSRequest(`type`: String, id: Option[String], payload: Option[GraphQLRequest])
+  object GraphQLWSRequest {
+    import io.circe._
+    import io.circe.generic.semiauto._
+
+    implicit val decodeGraphQLWSRequest: Decoder[GraphQLWSRequest] = deriveDecoder[GraphQLWSRequest]
+  }
+
+  def makeHttpService[R, E](
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean = false,
+    enableIntrospection: Boolean = true,
+    queryExecution: QueryExecution = QueryExecution.Parallel
+  ): HttpApp[R, HttpError] =
+    Http.collectM {
+      case req @ Method.POST -> _ =>
+        (for {
+          query           <- ZIO.fromEither(decode[GraphQLRequest](req.getBodyAsString.getOrElse("")))
+          queryWithTracing = req.headers
+                               .find(r => r.name == "apollo-federation-include-trace" && r.value == "ftv1")
```

Comment:
You can use the constants in `GraphQLRequest`

Reply:
Fixed in https://github.com/ghostdogpr/caliban/pull/867/commits/725953c48ebd6ad1eeeb0e84675c74e79dc2fa00